### PR TITLE
Lb wip2

### PR DESCRIPTION
**Pull request by mchalla**
The first commit addresses the code review comments. TODO
1. handle icmpv6 load balancing
2. verify the conntrack issue where the state lookup fails for some containers for reply traffic is fixed in latest master.
3. send packets to server using linux stack. (for some reason it did not work and i will debug it in latest master)

The second commit changes the addressing on go side to include <state>
<beef>:<beef>:<beef>:<beef>:<node>:<node>:<state>:<lxc>

### CODE REVIEW

File: bpf/lbmap/lb.go [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+	"github.com/codegangsta/cli"
+)
+
+const (
+	errInvalidArgument = -1
+	errIOFailure       = -2
+)
+
+const (
+	MaxLxc = 8
```
Comment (aanm) [inline bpf/lbmap/lb.go:31]: Only 8 LXCs per service? That low?

`MaxLxc` to `MaxLXC`

File: bpf/lbmap/lb.go [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+)
+
+// Lb1 : LbKey => LBValue
+// Lb2 : state => LBValue
+const (
+	ciliumMapLb1 = 1
```
Comment (aanm) [inline bpf/lbmap/lb.go]: rename `ciliumMapLb1` and `ciliumMapLb2` to a better name.

File: bpf/lbmap/lb.go [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+	ciliumMapLb2 = 2
+)
+
+type LbKey struct {
```
Comment (aanm) [inline bpf/lbmap/lb.go]: `LbKey` to `LBKey`

File: bpf/lbmap/lb.go [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+	vip   v6Addr
+	dport uint16
+	pad   uint16
+}
+
+type LbLxcPair struct {
```
Comment (aanm) [inline bpf/lbmap/lb.go]: `LbLxcPair` to `LBLXCPair`

File: bpf/lbmap/lb.go [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+	LxcId  uint16
```
Comment (aanm) [inline bpf/lbmap/lb.go]: `LxcId` to `LXCID`

File: bpf/lbmap/lb.go [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+	Port   uint16
+	NodeId uint32
```
Comment (aanm) [inline bpf/lbmap/lb.go]: `NodeId` to `NodeID`

File: bpf/lbmap/lb.go [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+}
+
+type LBValue struct {
+	vip       v6Addr
+	dport     uint16
+	state     uint16
+	lxc_count int32
```
Comment (aanm) [inline bpf/lbmap/lb.go]: `lxc_count` to `lxcCount`

File: bpf/lbmap/lb.go (hunk 1 above)
Comment (mchalla) [inline bpf/lbmap/lb.go:31]: We can change it later. It just takes memory on the bpf side, since its an array of MaxLXC per lb_value irrespective of how many you configure.

File: common/addressing/defaults.go [hunk 8]
```diff
@@ -24,16 +24,16 @@ var (
...
+	// lxc:                     beef:beef:beef:beef:<node>:<node>:<state>:<lxc>/128
 
 	// ClusterIPv6Mask represents the CIDR Mask for an entire cluster
 	ClusterIPv6Mask = net.CIDRMask(64, 128)
-	// StateIPv6Mask represents the CIDR Mask for the state porition
-	StateIPv6Mask = net.CIDRMask(80, 128)
 	// NodeIPv6Mask represents the CIDR Mask for the cilium node.
-	NodeIPv6Mask = net.CIDRMask(112, 128)
+	NodeIPv6Mask = net.CIDRMask(96, 128)
+	// StateIPv6Mask represents the CIDR Mask for the state porition
```
Comment (aanm) [inline common/addressing/defaults.go]: `porition` -> `position`

File: bpf/lbmap/lb.go (hunk 2 above)
Comment (mchalla) [inline bpf/lbmap/lb.go]: I could not come up with better names. They need to be consistent on both bpf and go side. Tell me what is the name u can think of based on the mapping comment I provided above.

File: common/addressing/ip.go [hunk 9]
```diff
@@ -52,17 +57,19 @@ func (ip CiliumIPv6) EndpointID() uint16 {
...
 // Returns true if IP is a valid IP for a container
 // To be valid must obey to the following rules:
-// - Node ID, bits from 112 to 120, must be different than 0
-// - Endpoint ID, bits from 120 to 128, must be different than 0
+// - Node ID, bits from 64 to 96, must be different than 0
+// - State, bits from 96 to 112, must be 0
+// - Endpoint ID, bits from 112 to 128, must be different than 0
 func (ip CiliumIPv6) ValidContainerIP() bool {
-	return ip.NodeID() != 0 && ip.EndpointID() != 0
+	return ip.NodeID() != 0 && ip.State() == 0 && ip.EndpointID() != 0
```
Comment (aanm) [inline common/addressing/ip.go:64]: I think a container IP can have State != 0 to be considered valid, what do you think @tgraf?

Comment (mchalla) [inline common/addressing/ip.go:64]: No it never will as seen by the container.

File: common/addressing/ip.go [hunk 10]
```diff
@@ -52,17 +57,19 @@ func (ip CiliumIPv6) EndpointID() uint16 {
...
 
 // Returns true if IP is a valid IP of a node
-// - Node ID, bits from 112 to 120, must be different than 0
-// - Endpoint ID, bits from 120 to 128, must be 0
+// - Node ID, bits from 64 to 96, must be different than 0
+// - State, bits from 96 to 112, must be 0
+// - Endpoint ID, bits from 112 to 128, must be 0
 func (ip *CiliumIPv6) ValidNodeIP() bool {
-	return ip.NodeID() != 0 && ip.EndpointID() == 0
+	return ip.NodeID() != 0 && ip.State() == 0 && ip.EndpointID() == 0
```
Comment (aanm) [inline common/addressing/ip.go:72]: I'm also not sure about this one, State could also be != 0. @tgraf

File: common/addressing/ip.go (hunk 9 above)
Comment (mchalla) [inline common/addressing/ip.go:64]: state is a transient thing, not something you should ever configure on the container ip.

File: common/addressing/ip.go (hunk 10 above)
Comment (mchalla) [inline common/addressing/ip.go:72]: same as above. Basically the state is configured as part of the map and only exists while the packet is in flight. Its never configured as part of node or container ip.

File: daemon/daemon/daemon.go [hunk 11]
```diff
@@ -229,7 +229,7 @@ func NewDaemon(c *Config) (*Daemon, error) {
 		return nil, fmt.Errorf("Configuration is nil")
 	}
 
-	ones, bits := addressing.NodeIPv6Mask.Size()
+	ones, bits := addressing.StateIPv6Mask.Size()
```
Comment (aanm) [inline daemon/daemon/daemon.go:236]: ~~Can you re-write this code?~~

~~Delete line 232, 233 and 234.~~

~~Replace `ipamSubnets := net.IPNet{IP: c.NodeAddress.IPv6Address.IP(), Mask: net.CIDRMask(maskPerIPAMType, bits)}` with `ipamSubnets := net.IPNet{IP: c.NodeAddress.IPv6Address.IP(), Mask: addressing.StateIPv6Mask}` and put it on a different commit because this might affect other stuff but probably won't.~~

Edit: I'll do it on a different pull request @aanm

Comment (mchalla) [inline daemon/daemon/daemon.go:236]: So you think the old code should instead have been Mask: addressing.NodeIpv6Mask ?

Comment (aanm) [inline daemon/daemon/daemon.go:236]: @mchalla don't worry I'll take a look at this later, for now `addressing.StateIPv6Mask.Size()` is the correct one ;-)

File: bpf/lbmap/lb.go (hunk 1 above)
Comment (mchalla) [inline bpf/lbmap/lb.go:31]: I will make this configurable later. That way only someone who needs many lxcs will need the additional memory.

File: bpf/lbmap/lb.go (hunk 2 above)
Comment (mchalla) [inline bpf/lbmap/lb.go]: may be I can call the first one CiliumMapLb and second one CiliumReverseMapLb, dont know if it conveys any additional meaning. Might as well leave it 1 and 2 :) but let me know and I will change it to something else.

Comment (aanm) [inline bpf/lbmap/lb.go]: Oh that's what they are supposed to be! Yes that helps a lot! Something like `lbMap` `lbReverseMap` you can leave `cilium` name out

File: bpf/lbmap/lb.go (hunk 5 above)
Comment (mchalla) [inline bpf/lbmap/lb.go]: What is the convention here ? all caps for LxcId ? and a diff convention for others ?

Comment (mchalla) [inline bpf/lbmap/lb.go]: I guess you mean LxcID like NodeID ?

File: common/addressing/ip.go (hunk 10 above)
Comment (tgraf) [inline common/addressing/ip.go:72]: Right, test looks correct. State should never be visible in user space.

File: bpf/bpf_lb.c [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+	__u16 sport;
+	__u32 hash;
+	struct lb_value *val;
+	int off = ETH_HLEN + sizeof(struct ipv6hdr);
+	int csum_off;
+	int ret = TC_ACT_OK;
+	__u8 nexthdr;
+	__be32 sum;
+
+	memcpy(&lb_ip.addr, lba, sizeof(lba));
```
Comment (tgraf) [inline bpf/bpf_lb.c]: Why is this needed? See other places like: https://github.com/noironetworks/cilium-net/blob/master/bpf/bpf_lxc.c#L87

File: bpf/bpf_lb.c [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+	if (likely(proto == __constant_htons(ETH_P_IPV6))) {
+		struct ipv6hdr *ip6 = data + ETH_HLEN;
+		union v6addr *dst = (union v6addr *) &ip6->daddr;
+
+		if (data + ETH_HLEN + sizeof(*ip6) > data_end) {
+			ret = DROP_INVALID;
+			goto error;
+		}
+
+		memcpy(&key.vip, dst, sizeof(*dst));
```
Comment (tgraf) [inline bpf/bpf_lb.c]: Can you use `ipv6_addr_copy()`?

File: bpf/bpf_lb.c [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+		csum_off = l4_checksum_offset(nexthdr);
+		if (unlikely(!csum_off))
+			goto error;
+
+		switch (nexthdr) {
+		case IPPROTO_TCP:
+			l4_load_port(skb, off + TCP_DPORT_OFF, &key.dport);
+			l4_load_port(skb, off + TCP_SPORT_OFF, &sport);
+			break;
+		case IPPROTO_UDP:
```
Comment (tgraf) [inline bpf/bpf_lb.c]: You can use the same execution block for TCP and UDP like done for the port mapping as the offsets for TCP and UDP are identical. Saves quite a bit of code.

File: bpf/bpf_lb.c [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+				if (key.dport != val->lxc[i].port) {
+					__u16 tmp = htons(val->lxc[i].port);
+					//cilium_trace(skb, DBG_GENERIC, val->lxc[i].port, 0);
+					switch (nexthdr) {
+					case IPPROTO_TCP:
+						l4_modify_port(skb, off + TCP_DPORT_OFF,
+							       csum_off, tmp,
+							       htons(key.dport));
+						break;
+					case IPPROTO_UDP:
```
Comment (tgraf) [inline bpf/bpf_lb.c]: Same here, TCP and UDP can be handled the same.

File: bpf/bpf_lb.c [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+		switch (nexthdr) {
+		case IPPROTO_TCP:
+			l4_load_port(skb, off + TCP_DPORT_OFF, &key.dport);
+			l4_load_port(skb, off + TCP_SPORT_OFF, &sport);
+			break;
+		case IPPROTO_UDP:
+			l4_load_port(skb, off + UDP_DPORT_OFF, &key.dport);
+			l4_load_port(skb, off + UDP_SPORT_OFF, &sport);
+			break;
+		default:
```
Comment (tgraf) [inline bpf/bpf_lb.c]: What about ICMPv6? At least add a FIXME to add it later.

File: bpf/bpf_lb.c [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+						l4_modify_port(skb, off + TCP_DPORT_OFF,
+							       csum_off, tmp,
+							       htons(key.dport));
+						break;
+					case IPPROTO_UDP:
+						l4_modify_port(skb, off + UDP_DPORT_OFF,
+							       csum_off, tmp,
+							       htons(key.dport));
+						break;
+					default:
```
Comment (tgraf) [inline bpf/bpf_lb.c]: This can never be reached. Simply drop out in the beginning for unknown nexthdr and then assume that stays valid. This blows up the code inside the unrolled loop.

File: bpf/bpf_lb.c [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+				if (i != which)
+					continue;
+
+				memcpy(&lb_ip.addr, server_prefix,
+				       sizeof(server_prefix));
+				ipv6_set_node_id(&lb_ip, val->lxc[i].node_id);
+				ipv6_set_state(&lb_ip, val->state);
+				ipv6_set_lxc_id(&lb_ip, val->lxc[i].lxc_id);
+				if (key.dport != val->lxc[i].port) {
+					__u16 tmp = htons(val->lxc[i].port);
```
Comment (tgraf) [inline bpf/bpf_lb.c:122]: Fine for now but the port number in the map should be stored in network byte order already so we don't have to convert in the fast path.

File: bpf/bpf_lb.c [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+				goto error;
+			}
+
+			eth_store_saddr(skb, lb_mac.addr, 0);
+			eth_store_daddr(skb, lxc_mac.addr, 0);
+
+			cilium_trace_capture(skb, DBG_CAPTURE_DELIVERY, SERVER_IFINDEX);
+			ret = redirect(SERVER_IFINDEX, 0);
+		} else {
+			//cilium_trace(skb, DBG_GENERIC, 100, 0);
```
Comment (tgraf) [inline bpf/bpf_lb.c]: Adding a new debug message for this doesn't hurt, it will be useful to know for troubleshoooting.

File: bpf/bpf_lxc.c [hunk 20]
```diff
@@ -142,6 +145,22 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 		return DROP_POLICY;
 	}
 
+	if (ret != CT_NEW) {
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: Why is the `ret != CT_NEW` check needed?

File: bpf/bpf_lxc.c [hunk 21]
```diff
@@ -142,6 +145,22 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
...
+		if (state) {
+			ret = lb_dsr_dnat(skb, state);
+			cilium_trace(skb, DBG_GENERIC, state, ret);
+			if (ret < 0)
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: `IS_ERR()`

File: bpf/bpf_lxc.c [hunk 22]
```diff
@@ -142,6 +145,22 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
...
+				return ret;
+		}
+	}
+
+	data = (void *) (long) skb->data;
+	data_end = (void *) (long) skb->data_end;
+	ip6 = data + ETH_HLEN;
+
+	if (data + sizeof(struct ipv6hdr) + ETH_HLEN > data_end)
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: Why is this needed? I don't see you access ip6 afterwards before the tail call.

File: bpf/bpf_lxc.c [hunk 23]
```diff
@@ -506,15 +527,26 @@ static inline int __inline__ ipv6_policy(struct __sk_buff *skb, int ifindex, __u
...
 
-	ret = ct_lookup6(&CT_MAP6, &tuple, skb, ETH_HLEN, SECLABEL, 1);
+	/*
+	 * derive state and zero it before doing conntrack lookup
+	 */
+	ipv6_load_daddr(skb, ETH_HLEN, &dip);
+	state = ipv6_derive_state(&dip);
+	if (state) {
+		ipv6_set_state(&dip, 0);
+		ipv6_store_daddr(skb, dip.addr, ETH_HLEN);
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: Checksum?

File: bpf/bpf_lxc.c [hunk 24]
```diff
@@ -506,15 +527,26 @@ static inline int __inline__ ipv6_policy(struct __sk_buff *skb, int ifindex, __u
 	tuple.nexthdr = ip6->nexthdr;
 	ipv6_addr_copy(&tuple.addr, (union v6addr *) &ip6->saddr);
 
-	ret = ct_lookup6(&CT_MAP6, &tuple, skb, ETH_HLEN, SECLABEL, 1);
+	/*
+	 * derive state and zero it before doing conntrack lookup
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: This comment does not make sense, you are not touching the address in the tuple.

File: bpf/bpf_netdev.c [hunk 25]
```diff
@@ -34,11 +34,6 @@ static inline int is_node_subnet(const union v6addr *dst, const union v6addr *no
 		tmp = dst->p2 - node_ip->p2;
 		if (!tmp) {
 			tmp = dst->p3 - node_ip->p3;
-			if (!tmp) {
-				__u32 a = ntohl(dst->p4);
-				__u32 b = ntohl(node_ip->p4);
-				tmp = (a & 0xFFFF0000) - (b & 0xFFFF0000);
-			}
```
Comment (tgraf) [inline bpf/bpf_netdev.c:41]: ACK

File: bpf/bpf_netdev.c [hunk 26]
```diff
@@ -52,11 +47,6 @@ static inline int matches_cluster_prefix(const union v6addr *addr, const union v
 	tmp = addr->p1 - prefix->p1;
 	if (!tmp) {
 		tmp = addr->p2 - prefix->p2;
-		if (!tmp) {
-			__u32 a = ntohl(addr->p3);
-			__u32 b = ntohl(prefix->p3);
-			tmp = (a & 0xFFFF0000) - (b & 0xFFFF0000);
-		}
```
Comment (tgraf) [inline bpf/bpf_netdev.c:59]: ACK. Matches `doc/model.md`

File: bpf/bpf_netdev.c [hunk 27]
```diff
@@ -52,11 +47,6 @@ static inline int matches_cluster_prefix(const union v6addr *addr, const union v
 	tmp = addr->p1 - prefix->p1;
 	if (!tmp) {
```
Comment (tgraf) [inline bpf/bpf_netdev.c]: Brackets can be removed after your change

File: bpf/bpf_netdev.c [hunk 28]
```diff
@@ -34,11 +34,6 @@ static inline int is_node_subnet(const union v6addr *dst, const union v6addr *no
 		tmp = dst->p2 - node_ip->p2;
 		if (!tmp) {
 			tmp = dst->p3 - node_ip->p3;
```
Comment (tgraf) [inline bpf/bpf_netdev.c]: Brackets can be removed after your change

File: bpf/lb_config.h [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+
+#define HANDLE_NS
+
+#define __LB_IPV6 { 0x20, 0x01, 0x0d, 0xb8, 0xaa, 0xaa, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x03 }
+#define __SERVER_PREFIX { 0xf0, 0x0d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0 }
+#define LB_IPV6 { .addr = __LB_IPV6 }
+
+#define ROUTER_IP __LB_IPV6
+#define SERVER_PREFIX __SERVER_PREFIX
+#define CLIENT_IFINDEX 3
```
Comment (tgraf) [inline bpf/lb_config.h]: What's this?

File: bpf/lb_config.h [hunk 30]
```diff
@@ -0,0 +1,23 @@
...
+#define __LB_CONFIG_H__
+
+#define HANDLE_NS
+
+#define __LB_IPV6 { 0x20, 0x01, 0x0d, 0xb8, 0xaa, 0xaa, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x03 }
+#define __SERVER_PREFIX { 0xf0, 0x0d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0 }
+#define LB_IPV6 { .addr = __LB_IPV6 }
+
+#define ROUTER_IP __LB_IPV6
+#define SERVER_PREFIX __SERVER_PREFIX
```
Comment (tgraf) [inline bpf/lb_config.h]: Why the indirection?

File: bpf/lb_config.h [hunk 31]
```diff
@@ -0,0 +1,23 @@
+#ifndef __LB_CONFIG_H__
+#define __LB_CONFIG_H__
+
+#define HANDLE_NS
+
+#define __LB_IPV6 { 0x20, 0x01, 0x0d, 0xb8, 0xaa, 0xaa, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x03 }
+#define __SERVER_PREFIX { 0xf0, 0x0d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0 }
+#define LB_IPV6 { .addr = __LB_IPV6 }
```
Comment (tgraf) [inline bpf/lb_config.h]: Unused?

File: bpf/lb_config.h [hunk 32]
```diff
@@ -0,0 +1,23 @@
...
+#define CLIENT_IFINDEX 3
+#define CLIENT_MAC { .addr = { 0x52, 0x54, 0x0, 0xa6, 0xf2, 0x22 } }
+#define NODE_MAC CLIENT_MAC
+#define SERVER_MAC { .addr = { 0x52, 0x54, 0x0, 0xa6, 0xf2, 0x22 } }
+
+/*
+ * This should be configured per lxc or derived via icmp6 neigh discovery
+ */
+#define LXC_MAC { .addr = { 0x52, 0x54, 0x0, 0xb3, 0x02, 0xe8 }}
+#define SERVER_IFINDEX 3
```
Comment (tgraf) [inline bpf/lb_config.h]: We can't merge this like this :-)

File: bpf/lbmap/lb.go (hunk 1 above)
Comment (tgraf) [inline bpf/lbmap/lb.go:31]: I think we all agree, A hardcoded 8 is fine to get started.

File: bpf/lbmap/lb.go (hunk 2 above)
Comment (tgraf) [inline bpf/lbmap/lb.go]: My suggestion: `lbServices` and `lbState`

File: bpf/lib/common.h [hunk 33]
```diff
@@ -157,6 +157,28 @@ struct ct_entry {
 	__u16 rx_closing:1,
 	      tx_closing:1,
 	      reserve:14;
+	__u16 state;
```
Comment (tgraf) [inline bpf/lib/common.h:162]: You need to extend `CtEntry` in `bpf/ctmap/ctmap.go` as well. I just noticed that the flags are missing as well. Allowing to dump the state via `cilium endponit ct` will help troubleshooting as well.

File: bpf/lib/ipv6.h [hunk 34]
```diff
@@ -142,9 +142,35 @@ static inline __u16 derive_lxc_id(union v6addr *addr)
 	return ntohl(addr->p4) & 0xFFFF;
 }
 
+static inline void ipv6_set_lxc_id(union v6addr *addr, __u16 lxc_id)
+{
+	__u32 p4 = ntohl(addr->p4);
+	p4 &= ~0xFFFF;
+	p4 |= lxc_id;
+	addr->p4 = htonl(p4);
```
Comment (tgraf) [inline bpf/lib/ipv6.h:220]: This could be rewritten to not require byte order conversion later.

File: bpf/lib/ipv6.h [hunk 35]
```diff
@@ -142,9 +142,35 @@ static inline __u16 derive_lxc_id(union v6addr *addr)
...
+{
+	return (ntohl(addr->p4) >> 16);
+}
+
+static inline void ipv6_set_state(union v6addr *addr, __u16 state)
+{
+	__u32 p4 = ntohl(addr->p4);
+	p4 &= 0xFFFF;
+	p4 |= (state << 16);
+	addr->p4 = htonl(p4);
```
Comment (tgraf) [inline bpf/lib/ipv6.h:234]: Same here.

File: bpf/lib/ipv6.h [hunk 36]
```diff
@@ -142,9 +142,35 @@ static inline __u16 derive_lxc_id(union v6addr *addr)
...
+}
+
 static inline __u32 ipv6_derive_node_id(union v6addr *addr)
 {
-	return (ntohl(addr->p3) & 0xFFFF) << 16 | (ntohl(addr->p4) >> 16);
+	return ntohl(addr->p3);
```
Comment (tgraf) [inline bpf/lib/ipv6.h:239]: Change is correct. Please also adjust the `Vagrantfile` since all addresses must be changed now.
`f00d::c0a8:210c:0` => `f00d::c0a8:210c:0:0`
`f00d::c0a8:210b:0/112` => `f00d::c0a8:210b:0:0/96`
And so on ...

File: bpf/lib/lb.h [hunk 37]
```diff
@@ -0,0 +1,55 @@
+#ifndef __LB_H_
+#define __LB_H_
+
+__BPF_MAP(cilium_lb2, BPF_MAP_TYPE_HASH, CILIUM_MAP_LB2, sizeof(__u16), sizeof(struct lb_value), PIN_GLOBAL_NS, 32);
```
Comment (tgraf) [inline bpf/lib/lb.h]: Make 32 a constant

File: bpf/lib/lb.h [hunk 38]
```diff
@@ -0,0 +1,55 @@
...
+
+	val = map_lookup_elem(&cilium_lb2, &state);
+	if (val != NULL) {
+		__u8 nexthdr;
+		union v6addr tmp = {}, sip;
+		__u16 sport;
+		int csum_off;
+		__be32 sum;
+
+		if (ipv6_load_nexthdr(skb, ETH_HLEN, &nexthdr) < 0 ||
```
Comment (tgraf) [inline bpf/lib/lb.h]: Pass the `tuple` into `lb_dsr_dnat` so you don't need to reload `nexthdr`

File: bpf/lib/lb.h [hunk 39]
```diff
@@ -0,0 +1,55 @@
...
+			break;
+
+		default:
+			return DROP_UNKNOWN_L4;
+		}
+
+		tmp.p1 = val->vip.p1;
+		tmp.p2 = val->vip.p2;
+		tmp.p3 = val->vip.p3;
+		tmp.p4 = val->vip.p4;
```
Comment (tgraf) [inline bpf/lib/lb.h:42]: Was this indirection needed to make the verifier happy?

File: bpf/lib/lb.h [hunk 40]
```diff
@@ -0,0 +1,55 @@
...
+		ipv6_store_saddr(skb, tmp.addr, ETH_HLEN);
```
Comment (tgraf) [inline bpf/lib/lb.h]: Handle possible error return and return `DROP_WRITE_ERROR`

File: bpf/lib/lb.h [hunk 41]
```diff
@@ -0,0 +1,55 @@
...
+
+		sum = csum_diff(sip.addr, 16, tmp.addr, 16, 0);
+		if (l4_csum_replace(skb, csum_off, 0, sum, BPF_F_PSEUDO_HDR) < 0)
+			return DROP_UNKNOWN_L4;
```
Comment (tgraf) [inline bpf/lib/lb.h]: Return `DROP_CSUM_L4`

File: bpf/lib/lb.h [hunk 42]
```diff
@@ -0,0 +1,55 @@
...
+		switch (nexthdr) {
+		case IPPROTO_TCP:
+			l4_load_port(skb, off + TCP_SPORT_OFF, &sport);
+			l4_modify_port(skb, off + TCP_SPORT_OFF, csum_off,
+				       htons(val->dport), sport);
+			break;
+
+		case IPPROTO_UDP:
+			l4_load_port(skb, off + UDP_SPORT_OFF, &sport);
+			l4_modify_port(skb, off + UDP_SPORT_OFF, csum_off,
```
Comment (tgraf) [inline bpf/lib/lb.h]: Handle possible error return

File: bpf/bpf_lb.c [hunk 43]
```diff
@@ -0,0 +1,163 @@
...
+				       sizeof(server_prefix));
+				ipv6_set_node_id(&lb_ip, val->lxc[i].node_id);
+				ipv6_set_state(&lb_ip, val->state);
+				ipv6_set_lxc_id(&lb_ip, val->lxc[i].lxc_id);
+				if (key.dport != val->lxc[i].port) {
+					__u16 tmp = htons(val->lxc[i].port);
+					//cilium_trace(skb, DBG_GENERIC, val->lxc[i].port, 0);
+					switch (nexthdr) {
+					case IPPROTO_TCP:
+						l4_modify_port(skb, off + TCP_DPORT_OFF,
```
Comment (tgraf) [inline bpf/bpf_lb.c]: Handle possible error return

File: bpf/bpf_lb.c [hunk 44]
```diff
@@ -0,0 +1,163 @@
...
+							       htons(key.dport));
+						break;
+					default:
+						ret = DROP_UNKNOWN_L4;
+						goto error;
+					}
+				}
+			}
+
+			ipv6_store_daddr(skb, lb_ip.addr, ETH_HLEN);
```
Comment (tgraf) [inline bpf/bpf_lb.c:140]: Handle possible error return

File: bpf/bpf_lb.c [hunk 45]
```diff
@@ -0,0 +1,163 @@
...
+
+			/* fixup csums */
+			sum = csum_diff(key.vip.addr, 16, lb_ip.addr, 16, 0);
+			if (l4_csum_replace(skb, csum_off, 0, sum, BPF_F_PSEUDO_HDR) < 0) {
+				ret = DROP_CSUM_L4;
+				goto error;
+			}
+
+			eth_store_saddr(skb, lb_mac.addr, 0);
```
Comment (tgraf) [inline bpf/bpf_lb.c:149]: Handle possible error return

File: bpf/lib/lb.h [hunk 46]
```diff
@@ -0,0 +1,55 @@
...
+		csum_off = l4_checksum_offset(nexthdr);
+		switch (nexthdr) {
+		case IPPROTO_TCP:
+			l4_load_port(skb, off + TCP_SPORT_OFF, &sport);
+			l4_modify_port(skb, off + TCP_SPORT_OFF, csum_off,
+				       htons(val->dport), sport);
+			break;
+
+		case IPPROTO_UDP:
+			l4_load_port(skb, off + UDP_SPORT_OFF, &sport);
```
Comment (tgraf) [inline bpf/lib/lb.h]: Handle possible error return

File: bpf/bpf_lb.c [hunk 47]
```diff
@@ -0,0 +1,163 @@
...
+#include "lib/common.h"
+#include "lib/maps.h"
+#include "lib/ipv6.h"
+#include "lib/icmp6.h"
+#include "lib/l4.h"
+#include "lib/eth.h"
+#include "lib/dbg.h"
+#include "lib/drop.h"
+
+__BPF_MAP(cilium_lb1, BPF_MAP_TYPE_HASH, CILIUM_MAP_LB1, sizeof(struct lb_key), sizeof(struct lb_value), PIN_GLOBAL_NS, 32);
```
Comment (tgraf) [inline bpf/bpf_lb.c]: This map doesn't need an ID, right? It's only accessed from `bpf_lb.c`
`__BPF_MAP(cilium_lb1, BPF_MAP_TYPE_HASH, 0, sizeof(struct lb_key), sizeof(struct lb_value), PIN_GLOBAL_NS, 32);`

**Comment by tgraf** [comment]
@mchalla Looks very good overall. Some comments inline. We need to make sure that all the existing automated tests and Vagrantfile still work after merging this.

File: bpf/lb_config.h (hunk 32 above)
Comment (mchalla) [inline bpf/lb_config.h]: This file is similar to the dummy node_config.h, I should have added a FIXME here. Its needed for compilation in the tree. It would be generated from the go code. There are couple of possibilities here, a load balancer with a single interface or two interfaces one facing client and other facing server. To simplify the vagrant setup I am using a single interface. During generation we need to pass this info to cilium so it compiles bpf_lb.c correctly. I am yet to implement this.

File: bpf/bpf_lb.c (hunk 47 above)
Comment (mchalla) [inline bpf/bpf_lb.c]: Thats correct. I did not know I could pass a 0, I will change it.

File: bpf/bpf_lxc.c (hunk 22 above)
Comment (mchalla) [inline bpf/bpf_lxc.c]: Its needed for the later ipv6_match_subnet_96 block.

File: bpf/bpf_lxc.c (hunk 24 above)
Comment (mchalla) [inline bpf/bpf_lxc.c]: Oh yes correct, the tuple uses the source address always. I forgot. Will remove it.

File: bpf/lb_config.h (hunk 31 above)
Comment (mchalla) [inline bpf/lb_config.h]: I will cleanup this file. I will move the generation of this file from cilium. I am thinking I will support single interface for now since thats what I am testing. But I will have to add a 2 interface load balancer.

File: bpf/lb_config.h (hunk 32 above)
Comment (tgraf) [inline bpf/lb_config.h]: OK, add a comment like in the other dummy header files then.

File: bpf/lb_config.h (hunk 31 above)
Comment (tgraf) [inline bpf/lb_config.h]: OK, it's fine to not ignore the generation to get started and require a handwritten header file for now.

File: bpf/bpf_lxc.c (hunk 22 above)
Comment (tgraf) [inline bpf/bpf_lxc.c]: Where is the change that requires it? `ipv6_local_delivery()` requires it which is why we check offets again in `bpf_lxc.c:170`.

Comment (mchalla) [inline bpf/bpf_lxc.c]: Not sure, we were debugging this issue in the chat session. Here are the details. Perhaps there is a better fix. I will remove it and see if it works since the code looks slightly different since then.
[debug.txt](https://github.com/noironetworks/cilium-net/files/409617/debug.txt)

Comment (mchalla) [inline bpf/bpf_lxc.c]: I do not have the old code. I think the 170 line was missing in the old code probably. I am basically duplicating that. So yep let me remove it and test it again.

File: bpf/bpf_lb.c (hunk 47 above)
Comment (mchalla) [inline bpf/bpf_lb.c]: Does starting cilium daemon on the lb create any of the maps by default ? If so it may need a different id. To use the logging framework (cilium monitor) we need to start the  daemon. Could that be an issue if we use 0 ?

Comment (mchalla) [inline bpf/bpf_lb.c]: It's per program it should be fine. I made it 0.

File: bpf/bpf_lb.c (hunk 17 above)
Comment (mchalla) [inline bpf/bpf_lb.c]: Yep I am already checking for that above the loop. Will remove it since the code will never change nexthdr.

File: bpf/lib/lb.h (hunk 39 above)
Comment (mchalla) [inline bpf/lib/lb.h:42]: Yes.
